modules/generator/processor/localblocks/processor.go

```diff
@@ -0,0 +1,402 @@
+package localblocks
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/google/uuid"
+	gen "github.com/grafana/tempo/modules/generator/processor"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+	"github.com/grafana/tempo/tempodb/wal"
+	"github.com/pkg/errors"
+)
+
+type Processor struct {
+	tenant  string
+	Cfg     Config
+	wal     *wal.WAL
+	closeCh chan struct{}
+
+	blocksMtx      sync.Mutex
+	headBlock      common.WALBlock
+	walBlocks      map[uuid.UUID]common.WALBlock
+	completeBlocks map[uuid.UUID]common.BackendBlock
+	lastCutTime    time.Time
+
+	liveTracesMtx sync.Mutex
+	liveTraces    *liveTraces
+}
+
+var _ gen.Processor = (*Processor)(nil)
+
+func New(cfg Config, tenant string, wal *wal.WAL) (*Processor, error) {
+
+	if wal == nil {
+		return nil, errors.New("local blocks processor requires traces wal")
+	}
+
+	p := &Processor{
+		Cfg:            cfg,
+		tenant:         tenant,
+		wal:            wal,
+		walBlocks:      map[uuid.UUID]common.WALBlock{},
+		completeBlocks: map[uuid.UUID]common.BackendBlock{},
+		liveTraces:     newLiveTraces(),
+		closeCh:        make(chan struct{}),
```

Comment:
Just thinking outloud, not blocking; do we prefer a close chan over a context cancellation?

Reply:
I like the chose channel for simplicity, but you make a good point that they have different behaviors.  Right now the processor would finish any in-progress flush or block conversion before shutting down, whereas context cancellation would immediately break out because it would be passed along to the i/o operations.